std/traits.d

```diff
@@ -8593,3 +8593,101 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+/++
+ + Determines if `F` is a strongly `pure` function. This is true when :
+ + $(UL
+ +   $(LI Each parameter is `immutable`, or can be converted automatically to `immutable`
+ +   (i.e. has no mutable indirections) and is passed by value)
+ +   $(LI The Return Type is `immutable` or can be impicitly converted to `immutable`)
+ + )
+ + Params:
+ + F = The function to be tested
+ +
+ + Returns:
+ + `true` if the function is strongly `pure`, `false` if otherwise.
+ + ++/
+template isStronglyPure(F...)
+if (F.length == 1 && isCallable!F)
+{
+    bool isStronglyPure()
+    {
+        if (!hasFunctionAttributes!(F, "pure"))
+            return false;
+
+        alias STC = ParameterStorageClass;
+        alias rt = ReturnType!F;
+        bool immutablert = true;
+
+        static foreach (i, param; Parameters!F)
+        {
+            if (is(param == immutable))
+            {}
+            else if (isRefType!param || hasAliasing!param
+                    || ParameterStorageClassTuple!F[i] == STC.ref_
+                    || ParameterStorageClassTuple!F[i] == STC.out_)
+                return false;
+            if (is(param : rt))
+                immutablert = false;
```

Comment:
It checks if any of the parameters are implicitly convertible to the return type. This is necessary because : 
```d
int[] foo(int[] a) pure
{
    return [1, 2, 3, 4, 5];
}

void main()
{
    int[] a = [1, 2];
    immutable b = foo(a); //fails, since b may become a slice of a 
    immutable c = foo([1, 2]); //ok
}
```
I decided its best to play safe and not call such functions strongly pure. Should they count as strongly pure?

Reply:
Thanks for clearing this up!